FTP

Inzilbeth opened the pull request.

Inzilbeth committed:
Some functionality added

Inzilbeth committed:
Update Program.cs

Inzilbeth committed:
Merge branch 'master' into homework4task1

Inzilbeth committed:
yml fix

Inzilbeth committed:
fix name and yml

Inzilbeth committed:
Update dotnet-core.yml

Inzilbeth committed:
Merge branch 'master' into homework4task1

Inzilbeth committed:
Implementation

Inzilbeth committed:
add ignored files

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,180 @@
+            {
+                var res = client.List("nonexistant").Result;
+
+                server.Stop();
```
Если client.List бросит исключение, то мы сюда не попадём уже и сервер никто не остановит.

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,180 @@
+                File.Delete(pathToSavedFile);
+
+                client.Stop();
+            });
```
Так оно до завершения теста не успеет выполниться. Стоило её подождать, а вообще, async-тесты тоже вполне себе должны работать.

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,180 @@
+        }
+
+        [Test]
+        public void SimoltaneousRequestsTest()
```
Sim**u**ltaneousRequestsTest

yurii-litvinov commented on Homework4/Task1/Task1Client/Client.cs:
```diff
@@ -0,0 +1,142 @@
+        /// <summary>
+        /// Simple demonstration of functionality.
+        /// </summary>
+        public async Task Start()
```
Это должно быть не здесь, это лучше в Main-е делать. Потому что по идее этот же клиент будет нужен в GUIшном варианте, для него CLI смысла не имеет

yurii-litvinov commented on Homework4/Task1/Task1Client/Client.cs:
```diff
@@ -0,0 +1,142 @@
+        /// <returns>List of (name, folder - true / file - false) </returns>
+        public async Task<(int, List<(string, bool)>)> List(string path)
+        {
+            var writer = new StreamWriter(client.GetStream()) { AutoFlush = true };
```
Они тоже IDisposable, надо их объявлять как using var

yurii-litvinov replied:
Напоминаю. Да, закрывающийся client закроет поток и освободит порт, но вообще надо давать всем IDisposable шанс исполнить свой Dispose, потому что у них, вообще говоря, могут быть свои какие-то внутренние ресурсы, которые надо финализировать.

yurii-litvinov commented on Homework4/Task1/Task1Client/Client.cs:
```diff
@@ -0,0 +1,142 @@
+                throw new FileNotFoundException();
+            }
+
+            using (var fileStream = new FileStream(pathTo + fileName, FileMode.CreateNew))
```
Можно было через обычный using var, чтобы без лишних скобок

yurii-litvinov commented on Homework4/Task1/Task1Client/Program.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        static async Task Main(string[] args)
+        {
+            var client = new Client("127.0.0.1", 9999);
```
Тут стоило получать IP/порт как args и как раз тут реализовать CLI (ведь Main --- это как раз то место, где стоит держать всё взаимодействие с пользователем, чтобы настоящие классы этим не занимались).

yurii-litvinov commented on Homework4/Task1/Task1Server/Program.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        static async Task Main(string[] args)
+        {
+            var server = new Server(9999);
```
Тут тоже, правильнее было получать порт как аргумент командной строки

yurii-litvinov commented on Homework4/Task1/Task1Server/Server.cs:
```diff
@@ -0,0 +1,118 @@
+
+                while (!cts.IsCancellationRequested)
+                {
+                    var client = await listener.AcceptTcpClientAsync();
```
TcpClient тоже IDisposable, но тут нельзя его просто как using var объявлять — он закроется, возможно, ещё до того, как потоку обработки запроса передадут управление. Надо сам клиент передавать в тот поток и там тут же брать на него using, чтобы по окончанию обработки запроса он корректно закрылся.

Inzilbeth committed:
fix

Inzilbeth committed:
Update FTPTests.cs

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,183 @@
+        }
+
+        [Test]
+        public async Task ListFunctionalityTest()
```
Видите, какой гнев вызвал async у GitHub Actions? Потому что async Вы написали, а await перед client.List нет, и блокируете поток с тестом до получения Result. Что kind of defeats the purpose.

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,183 @@
+            var response = client.List("Example").Result;
+
+            server.Stop();
+            client.Stop();
```
Я бы хотя бы для клиента using использовал --- а то вдруг что-то пойдёт не так и Stop не вызовется. А чтобы и сервер остановить, обернул бы всё в try/finally. Иначе из-за одного сломанного теста сервер будет занимать порт и другие тесты, возможно, не смогут выполниться (хотя им вроде всё равно к какому серверу подключаться, но всё же)

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,183 @@
+
+            client.Connect();
+
+            var response = client.List("Example\\testFolder").Result;
```
Тут тоже можно было вместо .Result написать await, тест сделать async, и это было бы идеологически правильнее, хотя по сути бы ничего не изменилось

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,183 @@
+                File.Delete(pathToSavedFile);
+            }
+
+            await Task.Run(async () =>
```
Мне кажется, это вообще не нужно. await Task.Run --- это "исполни следующий кусок кода в возможно другом потоке, без всякой видимой на то причины" --- ведь ничего параллельно не делается больше.

yurii-litvinov commented on Homework4/Task1/Task1Client/Client.cs:
```diff
@@ -0,0 +1,97 @@
+        /// <summary>
+        /// Simple demonstration of functionality.
+        /// </summary>
+        public async Task Start() => Connect();
```
Тут то же самое, async на самом деле не более чем "разрешает await-ы" внутри тела метода. Нет асинхронности --- нет радости.

yurii-litvinov commented on Homework4/Task1/Task1Client/Program.cs:
```diff
@@ -0,0 +1,67 @@
+            }
+            else if (args.Length != 0)
+            {
+                throw new ArgumentException();
```
Лучше какую-то более дружелюбную диагностику выводить. И с int.Parse тоже может не сложиться, и тоже по-хорошему надо внятно об этом пользователю сказать

yurii-litvinov commented on Homework4/Task1/Task1Client/Program.cs:
```diff
@@ -0,0 +1,67 @@
+
+                try
+                {
+                    if (input != null)
```
Надо фигурные скобки вокруг тела

yurii-litvinov commented on Homework4/Task1/Task1Client/Program.cs:
```diff
@@ -0,0 +1,67 @@
+                            {
+                                var res = await client.List(input);
+
+                                foreach (var e in res.Item2)
```
А ведь можно давать адекватные имена элементам кортежа, https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples#tuple-field-names

yurii-litvinov commented on Homework4/Task1/Task1Server/Program.cs:
```diff
@@ -0,0 +1,25 @@
+
+            if (args.Length == 1)
+            {
+                port = int.Parse(args[1]);
```
Мм, а это работает?

yurii-litvinov commented on Homework4/Task1/Task1Server/RequestHandler.cs:
```diff
@@ -0,0 +1,99 @@
+            }
+
+            var size = new FileInfo(path).Length;
+            await writer.WriteLineAsync($"{size}");
```
Согласно протоколу из условия, после размера должен быть пробел, а не перевод строки

yurii-litvinov commented on Homework4/Task1/Task1Server/Server.cs:
```diff
@@ -0,0 +1,114 @@
+        /// <summary>
+        /// Launch the server.
+        /// </summary>
+        public async Task Start()
```
Тут тоже печаль с async-ом

yurii-litvinov reviewed (changes requested):
Какая-то беда с async-ами и task-ами в тестах, см. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/. Правда, прочитайте и разберитесь, это, наверное, процентов 70 всего современного многопоточного программирования, причём не только в .NET

Inzilbeth committed:
fixes

Inzilbeth committed:
fix

Inzilbeth committed:
fix

Inzilbeth committed:
Update Program.cs

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,207 @@
+                File.Delete(pathToFile);
+            }
+
+            Task.Run(async () =>
```
Остался загадочный Task.Run в конце метода, который никто не await-ит

yurii-litvinov commented on Homework4/Task1/FTPTests/FTPTests.cs:
```diff
@@ -0,0 +1,207 @@
+
+                    Assert.IsTrue(File.Exists(pathToFile));
+
+                     (_, list) = client.List("Example\\testFolder").Result;
```
А тут ещё и .Result. То ли недоразбирались, то ли не хватило терпения везде исправить :)

yurii-litvinov commented on Homework4/Task1/Task1Client/Client.cs:
```diff
@@ -0,0 +1,97 @@
+        /// <summary>
+        /// Simple demonstration of functionality.
+        /// </summary>
+        public void Start() => Connect();
```
На самом деле, не думаю, что это вообще надо теперь

yurii-litvinov commented on Homework4/Task1/Task1Client/Client.cs:
```diff
@@ -0,0 +1,97 @@
+        /// </summary>
+        public void Connect()
+        {
+            client = new TcpClient(server, port);
```
Подключение суть длительная операция, и если уж мы дизайним красивую асинхронную абстракцию, стоило бы это делать в отдельном потоке, через await Task.Run. Или, что лучше, https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.tcpclient.connectasync?view=net-5.0

yurii-litvinov commented on Homework4/Task1/Task1Client/Client.cs:
```diff
@@ -0,0 +1,97 @@
+        public async Task<(int, List<(string, bool)>)> List(string path)
+        {
+            await using var writer = new StreamWriter(client.GetStream()) { AutoFlush = true };
+            using var reader = new StreamReader(client.GetStream());
```
А тут тоже можно await using вроде

Inzilbeth replied:
Компилятор ругается, нельзя вроде как

yurii-litvinov reviewed (approved):
Комментарии к коммитам ужасны :) Но вообще, стало лучше, с async/await теперь почти всё правильно.

Inzilbeth committed:
fixes

- changed client connection to async
- removed client's Start() method
- tests fixed
- changed "using ..." to "await using ... " where possible

yurii-litvinov commented on Homework4/Task1/Task1Client/Program.cs:
```diff
@@ -35,7 +35,7 @@ private static async Task Main(string[] args)
 
             var client = new Client(ip, port);
-            client.Start();
+            client.Connect();
```
Оно говорит, что тут теперь тоже надо await-ить, раз Connect стал асинхронным. Иначе мы можем начать посылать запросы ещё до того, как подключение установится.

yurii-litvinov reviewed (approved):
Ага, теперь в целом всё ок

Inzilbeth committed:
Merge branch 'master' into homework4task1
